revocation/revocation.go

```diff
@@ -56,26 +78,140 @@ func NewTimestamp(httpClient *http.Client) (Revocation, error) {
+		OCSPHTTPClient:   httpClient,
+		CRLHTTPClient:    httpClient,
+		CertChainPurpose: ocsp.PurposeTimestamping,
+	})
+}
+
+// NewWithOptions constructs a revocation object with the specified options
+func NewWithOptions(ctx context.Context, opts *Options) (Revocation, error) {
+	if ctx == nil {
+		return nil, errors.New("invalid input: a non-nil context must be specified")
+	}
+
+	if opts.OCSPHTTPClient == nil {
+		return nil, errors.New("invalid input: a non-nil OCSPHTTPClient must be specified")
+	}
+
+	if opts.CRLHTTPClient == nil {
+		return nil, errors.New("invalid input: a non-nil CRLHTTPClient must be specified")
+	}
+
+	switch opts.CertChainPurpose {
+	case ocsp.PurposeCodeSigning, ocsp.PurposeTimestamping:
+	default:
+		return nil, fmt.Errorf("unknown certificate chain purpose %v", opts.CertChainPurpose)
+	}
+
 	return &revocation{
-		httpClient:       httpClient,
-		certChainPurpose: ocsp.PurposeTimestamping,
+		ctx:              ctx,
+		ocspHTTPClient:   opts.OCSPHTTPClient,
+		crlHTTPClient:    opts.CRLHTTPClient,
+		certChainPurpose: opts.CertChainPurpose,
 	}, nil
 }
 
 // Validate checks the revocation status for a certificate chain using OCSP and
-// returns an array of CertRevocationResults that contain the results and any
-// errors that are encountered during the process
+// CRL if OCSP is not available. It returns an array of CertRevocationResults
+// that contain the results and any errors that are encountered during the
+// process.
+//
+// The certificate chain is expected to be in the order of leaf to root.
+//
+// This function tries OCSP and falls back to CRL when:
+// - OCSP is not supported by the certificate
+// - OCSP returns an unknown status
 //
-// TODO: add CRL support
-// https://github.com/notaryproject/notation-core-go/issues/125
+// When OCSP returns an unknown status, the function will try to check the
+// certificate status using CRL and return certificate result with an
+// result.OCSPFallbackError.
 func (r *revocation) Validate(certChain []*x509.Certificate, signingTime time.Time) ([]*result.CertRevocationResult, error) {
-	return ocsp.CheckStatus(ocsp.Options{
+	if len(certChain) == 0 {
+		return nil, result.InvalidChainError{Err: errors.New("chain does not contain any certificates")}
+	}
+
+	// Validate cert chain structure
+	// Since this is using authentic signing time, signing time may be zero.
+	// Thus, it is better to pass nil here than fail for a cert's NotBefore
+	// being after zero time
+	switch r.certChainPurpose {
+	case ocsp.PurposeCodeSigning:
+		if err := coreX509.ValidateCodeSigningCertChain(certChain, nil); err != nil {
+			return nil, result.InvalidChainError{Err: err}
+		}
+	case ocsp.PurposeTimestamping:
+		if err := coreX509.ValidateTimestampingCertChain(certChain); err != nil {
+			return nil, result.InvalidChainError{Err: err}
+		}
+	default:
+		return nil, result.InvalidChainError{Err: fmt.Errorf("unknown certificate chain purpose %v", r.certChainPurpose)}
+	}
```

Comment:
This part repeats what we already have under `ocsp.CheckStatus`. So instead of writing it twice, can we refactor it into a helper function that's shared by the two packages?

Reply:
Waiting for #215